crates/line_ending/src/lib.rs

```diff
@@ -0,0 +1,107 @@
+//! We maintain the invariant that all internal strings use `\n` as line separator.
+//! This module does line ending conversion and detection (so that we can
+//! convert back to `\r\n` on the way out as needed).
+
+use std::sync::LazyLock;
+
+use memchr::memmem;
+
+static FINDER: LazyLock<memmem::Finder> = LazyLock::new(|| memmem::Finder::new(b"\r\n"));
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum LineEnding {
+    ///  Line Feed only (\n), common on Linux and macOS as well as inside git repos
+    Lf,
+
+    /// Carriage Return + Line Feed characters (\r\n), common on Windows
+    Crlf,
+}
+
+pub fn infer(x: &str) -> LineEnding {
+    match FINDER.find(x.as_bytes()) {
+        // Saw `\r\n`
+        Some(_) => LineEnding::Crlf,
+        // No `\r\n`, or empty file
+        None => LineEnding::Lf,
+    }
+}
+
+/// Normalize line endings within a string
+///
+/// We replace `\r\n` with `\n` in-place, which doesn't break utf-8 encoding.
+/// While we *can* call `as_mut_vec` and do surgery on the live string
+/// directly, let's rather steal the contents of `x`. This makes the code
+/// safe even if a panic occurs.
+///
+/// # Source
+///
+/// ---
+/// authors = ["rust-analyzer team"]
+/// license = "MIT OR Apache-2.0"
+/// origin = "https://github.com/rust-lang/rust-analyzer/blob/master/crates/rust-analyzer/src/line_index.rs"
+/// ---
+pub fn normalize(x: String) -> String {
```

Comment:
I feel like it is more flexible to move towards providing attribution in doc comments

- It lets us structure the folders however we want, without needing a `rust-analyzer` folder
- It allows us to change function names and whatnot while still pointing to the original source